File: packages/ts/tests/integration/hermetic/extensions.test.ts
// Hermetic: first-party code runs for real; the mocked boundaries are
// `node:fs` probes and `node:module`'s `createRequire` (behind which both
// the napi binary and package resolution sit). Extension planning now lives
// in the core, so the fake core answers for it here; the real planner is
// covered by `tests/integration/binding/`.

import { existsSync, readFileSync, readdirSync, statSync } from "node:fs";
import { DirSQL } from "dirsql";
import { beforeEach, describe, expect, it, vi } from "vitest";

// Fake native core module, delivered through the mocked `createRequire` so
// the real `loadNativeCore()` / `getCore()` chain runs unmodified.
const { fakeCore } = vi.hoisted(() => ({
  fakeCore: {
    DirSQL: { openAsync: vi.fn() },
    planExtensionPath: vi.fn(),
    selectLoadable: vi.fn(),
    planConfigExtensions: vi.fn(),
  },
}));

vi.mock("node:fs", async (importOriginal) => ({
  ...(await importOriginal<typeof import("node:fs")>()),
  existsSync: vi.fn(),
  readFileSync: vi.fn(),
  readdirSync: vi.fn(),
  statSync: vi.fn(),
}));
vi.mock("node:module", async (importOriginal) => {
  const actual = await importOriginal<typeof import("node:module")>();
  // Called as `requirer(specifier)` by load-native-core: yield the fake core.
  const requirer = (_specifier: string) => fakeCore;
  // Read as `req.resolve` / `req.resolve.paths` by resolve-extension's
  // package prober: the fixture package resolves, everything else throws.
  requirer.resolve = (specifier: string): string => {
    if (specifier === "dirsql-testext-pkg/package.json") {
      return "/nm/dirsql-testext-pkg/package.json";
    }
    throw new Error(`Cannot find module '${specifier}'`);
  };
  requirer.resolve.paths = (): string[] => [];
  return {
    ...actual,
    createRequire: () => requirer as unknown as NodeJS.Require,
  };
});

const openAsync = fakeCore.DirSQL.openAsync;
const { planExtensionPath, selectLoadable, planConfigExtensions } = fakeCore;

beforeEach(() => {
  // Reset (not just clear) so per-test fs implementations never leak.
  vi.resetAllMocks();
  // The core plans a path-looking entry as a literal, verbatim.
  planExtensionPath.mockImplementation((path: string) => ({ path }));
  planConfigExtensions.mockReturnValue(null);
  openAsync.mockResolvedValue({
    query: vi.fn().mockResolvedValue([]),
    startWatcher: vi.fn().mockResolvedValue(undefined),
    pollEvents: vi.fn().mockResolvedValue([]),
  });
});

describe("extensions option (hermetic, #230/#299)", () => {
  it("defaults to null when no extensions are configured", async () => {
    const db = new DirSQL({ root: "/data" });
    await db.ready;
    expect(openAsync.mock.calls[0]?.[6]).toBeNull();
  });

  it("forwards literal paths verbatim with their entrypoint", async () => {
    const db = new DirSQL({
      root: "/data",
      extensions: [
        { path: "/ext/libvec.so", entrypoint: "sqlite3_vec_init" },
        { path: "./rel/spellfix.so" },
      ],
    });
    await db.ready;
    expect(openAsync.mock.calls[0]?.[6]).toEqual([
      { path: "/ext/libvec.so", entrypoint: "sqlite3_vec_init" },
      { path: "./rel/spellfix.so", entrypoint: undefined },
    ]);
    expect(existsSync).not.toHaveBeenCalled();
  });

  it("resolves a bare package name to the installed loadable", async () => {
    planExtensionPath.mockReturnValue({
      package: "dirsql-testext-pkg",
      shadow: "/cwd/dirsql-testext-pkg",
    });
    // No same-named local file shadows the package...
    vi.mocked(existsSync).mockReturnValue(false);
    // ...and the core picks the one loadable in the resolved package dir.
    vi.mocked(readdirSync).mockReturnValue([
      "libtestext.so",
    ] as unknown as ReturnType<typeof readdirSync>);
    selectLoadable.mockReturnValue("/nm/dirsql-testext-pkg/libtestext.so");

    const db = new DirSQL({
      root: "/data",
      extensions: [{ path: "dirsql-testext-pkg" }],
    });
    await db.ready;
    expect(openAsync.mock.calls[0]?.[6]).toEqual([
      { path: "/nm/dirsql-testext-pkg/libtestext.so", entrypoint: undefined },
    ]);
  });

  it("prefers a same-named local file over the installed package", async () => {
    planExtensionPath.mockReturnValue({
      package: "dirsql-testext-pkg",
      shadow: "/cwd/dirsql-testext-pkg",
    });
    vi.mocked(existsSync).mockReturnValue(true);
    vi.mocked(statSync).mockReturnValue({
      isFile: () => true,
    } as unknown as ReturnType<typeof statSync>);

    const db = new DirSQL({
      root: "/data",
      extensions: [{ path: "dirsql-testext-pkg" }],
    });
    await db.ready;
    expect(openAsync.mock.calls[0]?.[6]).toEqual([
      { path: "/cwd/dirsql-testext-pkg", entrypoint: undefined },
    ]);
  });

  it("resolves config [[dirsql.extension]] package names and suppresses the core's own loading (#313)", async () => {
    // The config file exists; nothing else does (in particular no local file
    // shadows the package name).
    vi.mocked(existsSync).mockImplementation((p) => p === "/cfg/.dirsql.toml");
    vi.mocked(readFileSync).mockReturnValue(
      '[[dirsql.extension]]\npath = "dirsql-testext-pkg"\n',
    );
    planConfigExtensions.mockReturnValue([
      { package: "dirsql-testext-pkg", shadow: "/cfg/dirsql-testext-pkg" },
    ]);
    vi.mocked(readdirSync).mockReturnValue([
      "libtestext.so",
    ] as unknown as ReturnType<typeof readdirSync>);
    selectLoadable.mockReturnValue("/nm/dirsql-testext-pkg/libtestext.so");

    const db = new DirSQL({ root: "/data", config: "/cfg/.dirsql.toml" });
    await db.ready;
    const call = openAsync.mock.calls[0];
    expect(call?.[6]).toEqual([
      { path: "/nm/dirsql-testext-pkg/libtestext.so", entrypoint: undefined },
    ]);
    expect(call?.[7]).toBe(true);
  });

  it("rejects ready when a bare package name is not installed", async () => {
    planExtensionPath.mockReturnValue({
      package: "not-installed-pkg",
      shadow: "/cwd/not-installed-pkg",
    });
    vi.mocked(existsSync).mockReturnValue(false);

    const db = new DirSQL({
      root: "/data",
      extensions: [{ path: "not-installed-pkg" }],
    });
    await expect(db.ready).rejects.toThrow(
      "could not resolve extension package 'not-installed-pkg': not installed",
    );
    expect(openAsync).not.toHaveBeenCalled();
  });
});

File: packages/python/tests/integration/hermetic/extensions_test.py
"""Hermetic integration tests for the ``extensions=`` constructor kwarg.

These exercise the SDK public API with the Rust core mocked and every
filesystem / import-machinery probe faked: programmatic extension entries
flow through ``dirsql.resolve_extension`` (run for real) and must reach the
core resolved. Only the third-party boundaries are patched -- the PyO3 core
class and the effectful stdlib probes (``os.path.isfile``,
``importlib.util.find_spec``, ``glob.glob``). Real extension *loading* (a
missing ``.so`` failing ready, a fixture cdylib actually registering a
function) is covered by ``tests/binding/``.
"""

import glob
import importlib.util
import os
from types import SimpleNamespace
from typing import ClassVar
from unittest.mock import mock_open, patch

import pytest

from dirsql import _async as async_mod


class _FakeRustDirSQL:
    """Test double for the PyO3 ``DirSQL`` class, recording constructor kwargs."""

    instances: ClassVar[list] = []

    def __init__(
        self,
        root=None,
        *,
        extensions=None,
        suppress_config_extensions=False,
        **kwargs,
    ):
        self.root = root
        self.extensions = extensions
        self.suppress_config_extensions = suppress_config_extensions
        self.kwargs = kwargs
        _FakeRustDirSQL.instances.append(self)


@pytest.fixture(autouse=True)
def _reset_instances():
    _FakeRustDirSQL.instances = []
    yield
    _FakeRustDirSQL.instances = []


@pytest.fixture
def mock_core():
    """Replace the Rust-backed ``_RustDirSQL`` alias in ``dirsql._async``."""
    with patch.object(async_mod, "_RustDirSQL", _FakeRustDirSQL):
        yield _FakeRustDirSQL


@pytest.fixture
def mock_cwd():
    """Pin the cwd the SDK resolves programmatic entries against."""
    with patch.object(os, "getcwd", return_value="/cwd") as m:
        yield m


@pytest.fixture
def mock_isfile():
    """Fake the bare-name local-file shadow probe (no local file by default)."""
    with patch.object(os.path, "isfile", return_value=False) as m:
        yield m


@pytest.fixture
def mock_find_spec():
    """Fake the installed-package lookup behind bare-name resolution."""
    with patch.object(importlib.util, "find_spec") as m:
        yield m


@pytest.fixture
def mock_glob():
    """Fake the loadable-file glob inside a resolved package dir."""
    with patch.object(glob, "glob") as m:
        yield m


@pytest.fixture
def mock_config_file():
    """Fake a ``.dirsql.toml`` whose ``[[dirsql.extension]]`` names a package."""
    data = b'[[dirsql.extension]]\npath = "sqlite_vec"\n'
    with patch("builtins.open", mock_open(read_data=data)) as m:
        yield m


def describe_extensions_kwarg():
    # Feature: DirSQL(extensions=[{path, entrypoint?}]) loads SQLite
    # extensions at startup. See docs/howto/load-extension.md and
    # packages/python/README.md.
    @pytest.mark.asyncio
    async def it_defaults_extensions_to_none(mock_core):
        db = async_mod.DirSQL("/root", tables=["t"])
        await db.ready()

        assert _FakeRustDirSQL.instances[0].extensions is None

    @pytest.mark.asyncio
    async def it_forwards_literal_paths_verbatim(mock_core, mock_cwd):
        db = async_mod.DirSQL(
            "/root",
            tables=["t"],
            extensions=[{"path": "/abs/libvec.so", "entrypoint": "sqlite3_vec_init"}],
        )
        await db.ready()

        assert _FakeRustDirSQL.instances[0].extensions == [
            {"path": "/abs/libvec.so", "entrypoint": "sqlite3_vec_init"}
        ]

    @pytest.mark.asyncio
    async def it_defaults_a_missing_entrypoint_to_none(mock_core, mock_cwd):
        db = async_mod.DirSQL(
            "/root", tables=["t"], extensions=[{"path": "./rel/libvec.so"}]
        )
        await db.ready()

        assert _FakeRustDirSQL.instances[0].extensions == [
            {"path": "./rel/libvec.so", "entrypoint": None}
        ]

    @pytest.mark.asyncio
    async def it_resolves_a_bare_package_name_to_the_installed_loadable(
        mock_core, mock_cwd, mock_isfile, mock_find_spec, mock_glob
    ):
        mock_find_spec.return_value = SimpleNamespace(
            submodule_search_locations=["/site-packages/sqlite_vec"], origin=None
        )
        mock_glob.return_value = ["/site-packages/sqlite_vec/vec0.so"]

        db = async_mod.DirSQL(
            "/root", tables=["t"], extensions=[{"path": "sqlite_vec"}]
        )
        await db.ready()

        assert _FakeRustDirSQL.instances[0].extensions == [
            {"path": "/site-packages/sqlite_vec/vec0.so", "entrypoint": None}
        ]

    @pytest.mark.asyncio
    async def it_prefers_a_same_named_local_file_over_the_package(
        mock_core, mock_cwd, mock_isfile
    ):
        mock_isfile.return_value = True

        db = async_mod.DirSQL(
            "/root", tables=["t"], extensions=[{"path": "sqlite_vec"}]
        )
        await db.ready()

        assert _FakeRustDirSQL.instances[0].extensions == [
            {"path": os.path.join("/cwd", "sqlite_vec"), "entrypoint": None}
        ]

    @pytest.mark.asyncio
    async def it_resolves_config_extension_package_names_and_suppresses_core_loading(
        mock_core, mock_cwd, mock_isfile, mock_find_spec, mock_glob, mock_config_file
    ):
        mock_isfile.side_effect = lambda p: p == "/cfg/.dirsql.toml"
        mock_find_spec.return_value = SimpleNamespace(
            submodule_search_locations=["/site-packages/sqlite_vec"], origin=None
        )
        mock_glob.return_value = ["/site-packages/sqlite_vec/vec0.so"]

        db = async_mod.DirSQL(config="/cfg/.dirsql.toml")
        await db.ready()

        inst = _FakeRustDirSQL.instances[0]
        assert inst.extensions == [
            {"path": "/site-packages/sqlite_vec/vec0.so", "entrypoint": None}
        ]
        assert inst.suppress_config_extensions is True

    @pytest.mark.asyncio
    async def it_rejects_ready_when_a_bare_package_name_is_not_installed(
        mock_core, mock_cwd, mock_isfile, mock_find_spec
    ):
        mock_find_spec.return_value = None

        db = async_mod.DirSQL("/root", tables=["t"], extensions=[{"path": "nope"}])
        with pytest.raises(ValueError, match="not installed"):
            await db.ready()

        # The resolution error surfaced before the core was ever constructed.
        assert _FakeRustDirSQL.instances == []
